### PR TITLE
Deflake `Bastion` controller integration test

### PR DESCRIPTION
**Pull request by shafeeqes**
**How to categorize this PR?**

/area testing
/kind flake

**What this PR does / why we need it**:
This PR introduces few minor improvements to solve few flakes:
Set the fakeClock time to `creationTimestamp.Time`. And also set the heartbeatTimestamp using `fakeClock.Now()` instead of `time.Now()`.
  The calculations inside the reconciler are made based on the `creationTimestamp`, the object is created a bit after we set fakeClock to `time.Now()`, and we set the `heartBeatTimestamp` ([and thus the expirationTimestamp](https://github.com/gardener/gardener/blob/5b3456739b23d2009b69e65b642892c507842a6a/pkg/registry/operations/bastion/strategy.go#L154)) using `time.Now()`, this causes the Bastion to be deleted immediately.
See: https://prow.gardener.cloud/view/gs/gardener-prow/logs/ci-gardener-integration/1594977484552540160#1:build-log.txt%3A49-55

Also during stress testing:
```
    [It] should delete Bastion if it's older than maxLifetime
      /Users/I545724/go/src/github.com/gardener/gardener/test/integration/controllermanager/bastion/bastion_test.go:242
    {"level":"debug","ts":"2022-11-25T14:33:58.698+0530","msg":"Requeuing Bastion","controller":"bastion","controllerGroup":"operations.gardener.cloud","controllerKind":"Bastion","Bastion":{"name":"test-5f324995","namespace":"bastion-controller-test-xmm92"},"namespace":"bastion-controller-test-xmm92","name":"test-5f324995","reconcileID":"c2e51e90-2a60-4560-9faa-8d972b647bd0","shoot":"bastion-controller-test-xmm92/test-5f324995","requeueAfter":"1.164058s"}
    {"level":"debug","ts":"2022-11-25T14:33:59.863+0530","msg":"Requeuing Bastion","controller":"bastion","controllerGroup":"operations.gardener.cloud","controllerKind":"Bastion","Bastion":{"name":"test-5f324995","namespace":"bastion-controller-test-xmm92"},"namespace":"bastion-controller-test-xmm92","name":"test-5f324995","reconcileID":"efbd39e7-d4a3-42d1-b5ec-dfabbf66cadd","shoot":"bastion-controller-test-xmm92/test-5f324995","requeueAfter":"1.164058s"}
    {"level":"debug","ts":"2022-11-25T14:34:01.028+0530","msg":"Requeuing Bastion","controller":"bastion","controllerGroup":"operations.gardener.cloud","controllerKind":"Bastion","Bastion":{"name":"test-5f324995","namespace":"bastion-controller-test-xmm92"},"namespace":"bastion-controller-test-xmm92","name":"test-5f324995","reconcileID":"aaac0148-88cb-42d2-9fb7-fa5d0f6b07db","shoot":"bastion-controller-test-xmm92/test-5f324995","requeueAfter":"1.164058s"}
    {"level":"debug","ts":"2022-11-25T14:34:02.194+0530","msg":"Requeuing Bastion","controller":"bastion","controllerGroup":"operations.gardener.cloud","controllerKind":"Bastion","Bastion":{"name":"test-5f324995","namespace":"bastion-controller-test-xmm92"},"namespace":"bastion-controller-test-xmm92","name":"test-5f324995","reconcileID":"dda7f9f8-49e3-4035-8007-b6c53c3288a7","shoot":"bastion-controller-test-xmm92/test-5f324995","requeueAfter":"1.164058s"}
    {"level":"debug","ts":"2022-11-25T14:34:03.359+0530","msg":"Requeuing Bastion","controller":"bastion","controllerGroup":"operations.gardener.cloud","controllerKind":"Bastion","Bastion":{"name":"test-5f324995","namespace":"bastion-controller-test-xmm92"},"namespace":"bastion-controller-test-xmm92","name":"test-5f324995","reconcileID":"22841b65-74fb-4a2d-9a20-b1488090483e","shoot":"bastion-controller-test-xmm92/test-5f324995","requeueAfter":"1.164058s"}
```
The creationTimestamp is again >1s than the fakeClock time, so even after stepping a second, the bastion is not deleted.

Also I am stepping 2 seconds more than the maxLifeTime, to prevent cases like the requeueAfter is some milliseconds.
```
      {"level":"debug","ts":"2022-11-25T14:36:10.387+0530","msg":"Requeuing Bastion","controller":"bastion","controllerGroup":"operations.gardener.cloud","controllerKind":"Bastion","Bastion":{"name":"test-5f324995","namespace":"bastion-controller-test-l85dc"},"namespace":"bastion-controller-test-l85dc","name":"test-5f324995","reconcileID":"ab73e9ce-644c-4702-b298-fa61cd17c253","shoot":"bastion-controller-test-l85dc/test-5f324995","requeueAfter":"377.093ms"}
      {"level":"debug","ts":"2022-11-25T14:36:10.765+0530","msg":"Requeuing Bastion","controller":"bastion","controllerGroup":"operations.gardener.cloud","controllerKind":"Bastion","Bastion":{"name":"test-5f324995","namespace":"bastion-controller-test-l85dc"},"namespace":"bastion-controller-test-l85dc","name":"test-5f324995","reconcileID":"7fac2c66-8743-45ec-aec4-2845a4956b65","shoot":"bastion-controller-test-l85dc/test-5f324995","requeueAfter":"377.093ms"}
      {"level":"debug","ts":"2022-11-25T14:36:11.143+0530","msg":"Requeuing Bastion","controller":"bastion","controllerGroup":"operations.gardener.cloud","controllerKind":"Bastion","Bastion":{"name":"test-5f324995","namespace":"bastion-controller-test-l85dc"},"namespace":"bastion-controller-test-l85dc","name":"test-5f324995","reconcileID":"e6dd04b0-ec13-43ad-8176-fb44901b5c38","shoot":"bastion-controller-test-l85dc/test-5f324995","requeueAfter":"377.093ms"}
      {"level":"debug","ts":"2022-11-25T14:36:11.522+0530","msg":"Requeuing Bastion","controller":"bastion","controllerGroup":"operations.gardener.cloud","controllerKind":"Bastion","Bastion":{"name":"test-5f324995","namespace":"bastion-controller-test-l85dc"},"namespace":"bastion-controller-test-l85dc","name":"test-5f324995","reconcileID":"ed8e16c8-fb8b-4191-80e0-429faf21ec55","shoot":"bastion-controller-test-l85dc/test-5f324995","requeueAfter":"377.093ms"}
      {"level":"debug","ts":"2022-11-25T14:36:11.900+0530","msg":"Requeuing Bastion","controller":"bastion","controllerGroup":"operations.gardener.cloud","controllerKind":"Bastion","Bastion":{"name":"test-5f324995","namespace":"bastion-controller-test-l85dc"},"namespace":"bastion-controller-test-l85dc","name":"test-5f324995","reconcileID":"566d2328-213a-4d80-9e0c-603c0512f3c1","shoot":"bastion-controller-test-l85dc/test-5f324995","requeueAfter":"377.093ms"}

  Waiting for:
      Deleting bastion because it reached its maximum lifetime
  In [It] at: /gardener/gardener/test/integration/controllermanager/bastion/bastion_test.go:248
```

Before:
```
$ stress -ignore "unable to grab random port | resource quota evaluation timed out" -p 64 ./test/integration/controllermanager/bastion/bastion.test
4m20s: 624 runs so far, 12 failures (1.92%)
4m25s: 646 runs so far, 12 failures (1.86%)
4m30s: 652 runs so far, 12 failures (1.84%)
4m35s: 668 runs so far, 12 failures (1.80%)
4m40s: 685 runs so far, 12 failures (1.75%)
4m45s: 710 runs so far, 12 failures (1.69%)
```

After:
```
$ stress -ignore "unable to grab random port | resource quota evaluation timed out" -p 64 ./test/integration/controllermanager/bastion/bastion.test
28m55s: 5054 runs so far, 0 failures
29m0s: 5072 runs so far, 0 failures
29m5s: 5085 runs so far, 0 failures
29m10s: 5100 runs so far, 0 failures
29m15s: 5116 runs so far, 0 failures
```

**Which issue(s) this PR fixes**:
Fixes https://prow.gardener.cloud/view/gs/gardener-prow/logs/ci-gardener-integration/1594977484552540160#1:build-log.txt%3A49-55

**Special notes for your reviewer**:
/cc @rfranzke 

**Release note**:

```other operator
NONE
```

### CODE REVIEW

**Comment by shafeeqes** [comment]
I actually saw failures with
```
      {"level":"debug","ts":"2022-11-25T15:00:11.461+0530","msg":"Requeuing Bastion","controller":"bastion","controllerGroup":"operations.gardener.cloud","controllerKind":"Bastion","Bastion":{"name":"test-9a221f5d","namespace":"bastion-controller-test-rg8rp"},"namespace":"bastion-controller-test-rg8rp","name":"test-9a221f5d","reconcileID":"bc2926ec-f709-4bba-a394-d3f14d7d6e00","shoot":"bastion-controller-test-rg8rp/test-9a221f5d","requeueAfter":"1s"}
      {"level":"debug","ts":"2022-11-25T15:00:12.462+0530","msg":"Requeuing Bastion","controller":"bastion","controllerGroup":"operations.gardener.cloud","controllerKind":"Bastion","Bastion":{"name":"test-9a221f5d","namespace":"bastion-controller-test-rg8rp"},"namespace":"bastion-controller-test-rg8rp","name":"test-9a221f5d","reconcileID":"10698f0d-456a-41bc-99ec-170eb7ef17f3","shoot":"bastion-controller-test-rg8rp/test-9a221f5d","requeueAfter":"1s"}
      {"level":"debug","ts":"2022-11-25T15:00:13.463+0530","msg":"Requeuing Bastion","controller":"bastion","controllerGroup":"operations.gardener.cloud","controllerKind":"Bastion","Bastion":{"name":"test-9a221f5d","namespace":"bastion-controller-test-rg8rp"},"namespace":"bastion-controller-test-rg8rp","name":"test-9a221f5d","reconcileID":"66f3a1a7-061e-4b35-9829-34aaee828635","shoot":"bastion-controller-test-rg8rp/test-9a221f5d","requeueAfter":"-1s"}

  Waiting for:
      Deleting expired bastion
  In [It] at: /Users/*/go/src/github.com/gardener/gardener/test/integration/controllermanager/bastion/bastion_test.go:239
```
with `requeueAfter: -1s`, but I can't reproduce it now, should we set these negative `time.Duration` to zero in the reconciler?

**Comment by shafeeqes** [comment]
Not related to this PR,
/test pull-gardener-integration

**Review by timuthy — COMMENTED** [review]
Looks good!

About your comment:

> with requeueAfter: -1s, but I can't reproduce it now, should we set these negative time.Duration to zero in the reconciler?

Having a look at the [requeueAfter calculation](https://github.com/gardener/gardener/blob/master/pkg/controllermanager/controller/bastion/reconciler.go#L107) I think we can handle negative durations although it's unlikely that we'll run into such cases. However, it should be something >0 seconds.

**Comment by shafeeqes** [comment]
> Looks good!
> 
> About your comment:
> 
> > with requeueAfter: -1s, but I can't reproduce it now, should we set these negative time.Duration to zero in the reconciler?
> 
> Having a look at the [requeueAfter calculation](https://github.com/gardener/gardener/blob/master/pkg/controllermanager/controller/bastion/reconciler.go#L107) I think we can handle negative durations although it's unlikely that we'll run into such cases. However, it should be something >0 seconds.

I just added https://github.com/gardener/gardener/pull/7080/commits/47f2001ea1b6163e8b235374b0fea1ad9af3c308

File: pkg/controllermanager/controller/bastion/reconciler.go [hunk 1]
```diff
@@ -112,6 +112,10 @@ func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
 		}
 	}
 
+	if requeueAfter < 0 {
+		requeueAfter = 0
```
Comment (timuthy) [inline pkg/controllermanager/controller/bastion/reconciler.go]: Hm, don't we want to requeue? Then it must be >0.

Comment (shafeeqes) [reply pkg/controllermanager/controller/bastion/reconciler.go]: True. How about returning an error so that it's requeued immediately.

**Review by timuthy — COMMENTED** [review]
Thank you!
/lgtm

Will approve later to give other reviewers a chance to take a second look.

**Comment by shafeeqes** [comment]
> @shafeeqes: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> 
> Test name	Commit	Details	Required	Rerun command
> pull-gardener-e2e-kind-ha-multi-zone	[101760e](https://github.com/gardener/gardener/commit/101760ed8bf2a4b76ff99ecca75b70741c38d0f2)	[link](https://prow.gardener.cloud/view/gs/gardener-prow/pr-logs/pull/gardener_gardener/7080/pull-gardener-e2e-kind-ha-multi-zone/1597140664330489856)	true	`/test pull-gardener-e2e-kind-ha-multi-zone`
> pull-gardener-e2e-kind	[101760e](https://github.com/gardener/gardener/commit/101760ed8bf2a4b76ff99ecca75b70741c38d0f2)	[link](https://prow.gardener.cloud/view/gs/gardener-prow/pr-logs/pull/gardener_gardener/7080/pull-gardener-e2e-kind/1597140664703782912)	true	`/test pull-gardener-e2e-kind`
> [Full PR test history](https://prow.gardener.cloud/pr-history?org=gardener&repo=gardener&pr=7080). [Your PR dashboard](https://prow.gardener.cloud/pr?query=is%3Apr%20state%3Aopen%20author%3Ashafeeqes). [Command help for this repository](https://prow.gardener.cloud/command-help?repo=gardener%2Fgardener). Please help us cut down on flakes by linking this test failure to an [open flake report](https://github.com/gardener/gardener/issues?q=is:issue+is:open+label:kind/flake) or filing a [new flake report](https://github.com/gardener/gardener/issues/new?template=flaking-test.md&title=%5BFlaky+Test%5D+FLAKING+TEST%2FSUITE) if you can't find an existing one. Also see our [testing guideline](https://github.com/gardener/gardener/blob/master/docs/development/testing.md) for how to avoid and hunt flakes.


Probably #7064 
/test pull-gardener-e2e-kind-ha-multi-zone

**Comment by shafeeqes** [comment]
> link

Probably https://github.com/gardener/gardener/issues/7064
/test pull-gardener-e2e-kind-ha-multi-zone

**Pull request merged**